File: next/dweb-browser/Helper/src/SortHelper.cs

namespace DwebBrowser.Helper;

public class ComparableWrapperBuilder<T>
{
    public Func<T, Dictionary<string, int>> GetScore { get; init; }
    public ComparableWrapperBuilder(Func<T, Dictionary<string, int>> getScore)
    {
        GetScore = getScore;
    }

    public ComparableWrapper<T> Build(T value) => new(value, GetScore);
}

public class ComparableWrapper<T> : IComparable<ComparableWrapper<T>>
{
    public T Value { get; init; }
    public Func<T, Dictionary<string, int>> GetScore { get; init; }

    public ComparableWrapper(T value, Func<T, Dictionary<string, int>> getScore)
    {
        Value = value;
        GetScore = getScore;
    }

    public static ComparableWrapperBuilder<T> Builder(Func<T, Dictionary<string, int>> getScore) => new(getScore);

    private Dictionary<string, int>? score = null;

    public Dictionary<string, int> Score => score ??= GetScore(Value);

    private HashSet<string> ScopeKeys => Score.Keys.ToHashSet();

    public int CompareTo(ComparableWrapper<T> other)
    {
        var aScore = score;
        var bScore = other.score;

        foreach (var key in ScopeKeys)
        {
            var aValue = aScore.GetValueOrDefault(key);
            var bValue = bScore.GetValueOrDefault(key);

            if (aValue != bValue)
            {
                return aValue - bValue;
            }
        }

        return 0;
    }
}

public static class EnumComparable
{
    public static int EnumToComparable<T>(T enumValue, List<T> enumList) => enumList.IndexOf(enumValue);

    public static List<int> EnumToComparable<T>(IEnumerable<T> enumValues, List<T> enumList)
    {
        var sorts = new List<int>();
        foreach (var value in enumValues)
        {
            sorts.Add(enumList.IndexOf(value));
        }

        return sorts.OrderBy(it => it).ToList();
    }
}

File: next/kmp/helper/src/commonMain/kotlin/org/dweb_browser/helper/sortHelper.kt
package org.dweb_browser.helper

public class ComparableWrapper<T>(
  public val value: T,
  public val getScore: (T) -> Map<String, Int>,
) :
  Comparable<ComparableWrapper<T>> {
  public companion object {
    public fun <T> builder(getScore: (T) -> Map<String, Int>): ComparableWrapperBuilder<T> =
      ComparableWrapperBuilder(getScore)

    public class ComparableWrapperBuilder<T>(private val getScore: (T) -> Map<String, Int>) {
      public fun build(value: T): ComparableWrapper<T> = ComparableWrapper(value, getScore)
    }
  }

  private var _score: Map<String, Int>? = null
  public val score: Map<String, Int>
    get() {
      if (_score == null) {
        _score = getScore(value)
      }
      return _score!!
    }

  private val _scopeKeys
    get() = score.keys

  /**
   * 等价于 `(this as a) - b`
   */
  override fun compareTo(other: ComparableWrapper<T>): Int {
    val aScore = score
    val bScore = other.score
    _scopeKeys.forEach { key ->
      val aValue = aScore[key] ?: 0
      val bValue = bScore[key] ?: 0
      if (aValue != bValue) {
        return aValue - bValue
      }
    }
    return 0
  }
}

public fun <T> enumToComparable(enumValue: T, enumList: List<T>): Int = enumList.indexOf(enumValue)
public fun <T> enumToComparable(enumValues: Iterable<T>, enumList: List<T>): List<Int> =
  enumValues.map { enumList.indexOf(it) }.sorted()
